### PR TITLE
Function fixes

### PR DESCRIPTION
**Pull request by gabi-250**
Function parameters are now local variables. Before, the parameters were permanently in scope (they were stored as global variables).

### CODE REVIEW

File: state.py [hunk 1]
```diff
@@ -7,8 +7,47 @@
 class NoSuchVariableException(Exception):
     pass
 
+class State(object):
```
Comment (vext01) [inline state.py]: I wonder why we don't just call this class `Frame`, since that's all it is, right?

**Comment by vext01** [comment]
So does your LLVM (and your interpreter) support global variables?

**Comment by vext01** [comment]
Sorry for the late response on this by the way.

**Comment by vext01** [comment]
There's a fair amount of special casing for the frame of the main function, which somewhat surprises me. Is the main frame not a frame like any other which just happens to have `argc` and `argv` variables, or is there more to it?

File: state.py (hunk 1 above)
Comment (gabi-250) [reply state.py]: Yep, you're right, it would make more sense to call it `Frame`.

**Comment by gabi-250** [comment]
It does support global variables, although I should probably write a test to see if they actually work as they should. I'm pretty confident they are correctly handled since all of my tests include globals. 

The main frame is a regular frame. The only reason it seems to be special cased is because I have kind of an odd way of managing function calls. Whenever a function is called (such as main), a brand new stack frame is created and passed to the instance of `Interpreter` that interprets it (I create a new interpreter for each call). Note that the new stack frame will map the parameters of the function to the value of the arguments the function is called with. I chose this approach because the only way to obtain the arguments of a function is by examining the `call` instruction and its own arguments. I cannot obtain the value that is supposed to be assigned to the parameters from the basic blocks of the function. So, to solve this problem, I identify the arguments before interpreting the function. The `main` function is no, the interpreter needs to know the value of `argc` and `argv` ahead of time.

I hope it makes sense. I know it's not the most efficient approach, but it's the most straightforward one I could think of.

**Comment by vext01** [comment]
I'm not 100% sure I understand. Are you coming to the next thrash? If so, perhaps you could explain with an example? Maybe we can come up with something more efficient...

The stuff we are talking about above is separate from the functionality you are implementing int the, so I won't block the merge.

What would be nice though, is a test to check the local variable scoping order works. I suppose a local variable shadows a global one?

**Comment by gabi-250** [comment]
When is the next thrash? Even if I come, I'm not sure I'll be able to stay for the whole thing.
Yep, local variables are checked first. I could write a test for the variables.

**Comment by vext01** [comment]
Looks good for now. Do you want to squash?

**Comment by gabi-250** [comment]
Sure. I'll write a test for scope in another branch then.

**Comment by gabi-250** [comment]
Done.

**Comment by gabi-250** [comment]
I've raised an issue regarding the lack of tests for variable scoping. Can we merge this branch?

**Pull request merged**